internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,333 @@
+	return selector.Kind == reference.Kind && selector.APIVersion == reference.APIVersion
+}
+
+var validOps = sets.NewString("add", "replace", "remove")
+
+func validateJSONPatches(jsonPatches []clusterv1.JSONPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	variableSet := getClusterClassVariablesMap(variables)
+
+	for i, jsonPatch := range jsonPatches {
+		if !validOps.Has(jsonPatch.Op) {
+			allErrs = append(allErrs,
+				field.NotSupported(
+					path.Index(i).Child("op"),
+					jsonPatch,
+					validOps.List(),
+				))
+		}
+
+		if !strings.HasPrefix(jsonPatch.Path, "/spec/") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch path must start with \"/spec/\".",
+				))
+		}
+
+		// return early if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+		if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch must define one of value or valueFrom.",
+				))
+		}
+
+		if jsonPatch.Value != nil && jsonPatch.ValueFrom != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch can not define both value and valueFrom.",
+				))
+		}
+
+		// Attempt to marshal the JSON to discover if it  is valid.
+		if jsonPatch.Value != nil {
+			if err := json.Unmarshal(jsonPatch.Value.Raw, &struct{}{}); err != nil {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Index(i),
+						jsonPatch.Value,
+						"jsonPatch Value is invalid JSON.",
+					))
+			}
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template == nil && jsonPatch.ValueFrom.Variable == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i).Child("valueFrom"),
+					jsonPatch,
+					"valueFrom must set either \"/template or variable/\".",
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil && jsonPatch.ValueFrom.Variable != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i).Child("valueFrom"),
+					jsonPatch,
+					"valueFrom can not set both \"/Template and variable/\".",
+				))
+		}
+
+		// Validate that array access is only prepend or append for add and not allowed for replace.
+		allErrs = append(allErrs,
+			validateIndexAccess(jsonPatch, path.Index(i).Child("path"))...,
+		)
+
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil {
+			// Error if template can not be parsed.
+			_, err := template.New("tpl").Parse(*jsonPatch.ValueFrom.Template)
+			if err != nil {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Index(i).Child("valueFrom", "template"),
+						jsonPatch.ValueFrom.Template,
+						fmt.Sprintf("Template can not be parsed. Error %s", err),
```

Comment:
```suggestion
						fmt.Sprintf("template can not be parsed, err: %v", err),
```
I think errors have to be formatted with `%v` (it's using `Error()` not `String()`)
(lower case because it's essentially an error string not a full sentence and the field name is lower case `template`)